### ISSUE TITLE
HLSL: calling function with wrong argument type (flatten loses const qualifier)

### ISSUE THREAD

**Issue by xxxbxxx** [KhronosGroup/glslang#953]
hello,

as noticed by the validtion layer
[SC] Code 5 : SPIR-V module not valid: OpFunctionCall Argument '407's type does not match Function '19's parameter type.

I think that
 - At the function declaration site, the flattening of the arguments discards the const qualifier.
 - While the caller does not.
 

```
struct S1
{
	float4 a;
	float b;
};

struct S2 
{
	Texture1D<float>	t;
	float			f;
	S1			s;
	uint			i;
};

struct S3
{
	float4			v;
	float			f;
	uint			i;
};


S3 func(in const S2 _In)
{
	const S1 s = _In.s;

	S3 Out;
	Out.v = s.a;
	Out.i = _In.i;
	Out.f = _In.f;

	return Out;
}


struct PS_INPUT {
	float4 	p : SV_Position;
};
struct PS_OUTPUT {
	float4 c : SV_Target0;
};

cbuffer C1
{
	S1	uni1;
};


Texture1D<float> T : register(t0);

PS_OUTPUT main(const PS_INPUT v)
{
	PS_OUTPUT	Output;
	Output.c	= float4(1,0,0,1);

	S2 	s;
	s.t = T;
	s.f = v.p.x;
	s.s = uni1;
	s.i = 1;
	
	S3 r = func(s);

	Output.c = r.v;	
	return Output;
}

```

**Comment by xxxbxxx** [comment]
This hack fixes the issue:

```
diff --git a/hlsl/hlslParseHelper.cpp b/hlsl/hlslParseHelper.cpp
index 29543d82..38b30af5 100755
--- a/hlsl/hlslParseHelper.cpp
+++ b/hlsl/hlslParseHelper.cpp
@@ -1761,8 +1761,10 @@ TIntermAggregate* HlslParseContext::handleFunctionDefinition(const TSourceLoc& l
                 const TTypeList* structure = variable->getType().getStruct();
                 for (int mem = 0; mem < (int)structure->size(); ++mem) {
                     initFlattening();
+                    TType dereferencedType(variable->getType(), mem);
+                    dereferencedType.getQualifier().storage = variable->getType().getQualifier().storage;
                     paramNodes = intermediate.growAggregate(paramNodes,
-                                                            flattenAccess(variable->getUniqueId(), mem, *(*structure)[mem].type),
+                                                            flattenAccess(variable->getUniqueId(), mem, dereferencedType),
                                                             loc);
                     finalizeFlattening();
                 }
```

**Comment by rohith10** [comment]
I think I ran into a weird mix of this and #944 - I had a function that was taking in a uniform/const qualified int2, and the call to the function was being made with a scalar int literal. SPIR-V emitted by glslang was trying to use the scalar as argument for the OpFunctionCall, which is obviously invalid (and causing the validation failure above).

When I removed the uniform qualifier from the source, correct SPIR-V was generated. Will dig deeper when I get a chance.
